object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -279,6 +303,248 @@ private static String getExactStringValue(Tree tree, ValueAnnotatedTypeFactory f
+              reportAlwaysCallErrors(
+                  assign,
+                  storeAfter,
+                  (last instanceof AssignmentNode) ? getAnnotatedType(last.getTree()) : null);
+            }
+
+            toRemove.add(assign);
+          }
+        }
+
+        newDefs.removeAll(toRemove);
+        propagate(Pair.of(succ, newDefs), visited, worklist);
+      }
+    }
+  }
+
+  /**
+   * it inputs a set of AssignmentNodes(defs) and a LocalVariableNode(localVariableNode), if there
+   * is an AssignmentNode in defs that its lhs is equal to localVarNode, then it returns that
+   * AssignmentNode, otherwise it returns null.
+   *
+   * @param defs
+   * @param node
+   * @return
+   */
+  private Pair<LocalVariableNode, Tree> getAssignmentTreeOfVar(
+      Set<Pair<LocalVariableNode, Tree>> defs, LocalVariableNode node) {
+    return defs.stream()
+        .filter(assign -> assign.first.getElement().equals(node.getElement()))
+        .findAny()
+        .orElse(null);
+  }
+
+  private boolean isVarInDefs(Set<Pair<LocalVariableNode, Tree>> defs, LocalVariableNode node) {
+    return defs.stream()
+        .map(assign -> ((assign.first).getElement()))
+        .anyMatch(elem -> elem.equals(node.getElement()));
+  }
+
+  /**
+   * If cur is Conditional block, then it returns a list of two successor blocks contains then block
+   * and else block. If cur is instance of SingleSuccessorBlock then it returns a list of one block.
+   * Otherwise, it throws an assertion error at runtime.
+   *
+   * @param cur
+   * @return list of successor blocks
+   */
+  private List<BlockImpl> getSuccessors(BlockImpl cur) {
+    List<BlockImpl> successorBlock = new ArrayList<>();
+
+    if (cur.getType() == Block.BlockType.CONDITIONAL_BLOCK) {
+
+      ConditionalBlock ccur = (ConditionalBlock) cur;
+
+      successorBlock.add((BlockImpl) ccur.getThenSuccessor());
+      successorBlock.add((BlockImpl) ccur.getElseSuccessor());
+
+    } else {
+      assert cur instanceof SingleSuccessorBlock;
+
+      Block b = ((SingleSuccessorBlock) cur).getSuccessor();
+      if (b != null) {
+        successorBlock.add((BlockImpl) b);
+      }
+    }
+    return successorBlock;
+  }
+
+  /**
+   * If the input block is Regular, then it returns a list of nodes exist in the content. If it's an
+   * Exception_Block, then it returns the corresponding node that causes exception. Otherwise, it
+   * returns an emptyList.
+   *
+   * @param block
+   * @return List of Nodes
+   */
+  private List<Node> getBlockNodes(Block block) {
+    List<Node> blockNodes = new ArrayList<>();
+
+    switch (block.getType()) {
+      case REGULAR_BLOCK:
+        blockNodes = ((RegularBlockImpl) block).getContents();
+        break;
+
+      case EXCEPTION_BLOCK:
+        blockNodes.add(((ExceptionBlockImpl) block).getNode());
+    }
+    return blockNodes;
+  }
+
+  /**
+   * It updates visited and worklist if the input state has not been visited yet
```

Comment:
Nit: the style of this comment is a bit confusing. I would rewrite it as:

`Updates {@code vistited} and {@code worklist} if the input {@code state} has not been visited yet.`